### ISSUE TITLE
Fish doesn't propagate SIGHUP from kitty terminal emulator exiting, causing children processes to keep running and consume resources

### ISSUE THREAD

**Issue by ronjouch** [fish-shell/fish-shell#6012]
This is a follow-up of https://github.com/kovidgoyal/kitty/issues/1863 , where I'm complaining about my terminal emulator Kitty. According to Kitty's maintainer, Kitty is doing the right thing:

> *"When a window is closed, the child program running in it is sent a `SIGHUP` signal. It is up to the child process how it handles that signal. Most terminal programs such as shells will quit on getting it and send it to all their child processes as well."*

, so I'm chiming here.

**Scenario**:

1. Start `kitty fish`
2. Launch a child process (e.g. `tig` in the below screenshot)
3. Exit `kitty`, using my window manager Close button.

**Expected**: Kitty sends a `SIGHUP`, fish handles it, fish terminates children processes, fish exits.

**Actual**: Kitty sends a `SIGHUP`, fish doesn't exit, fish doesn't terminate children processes.

Also,

- I couldn't reproduce the problem in `gnome-terminal -- fish`. When exiting gnome-term, fish seems to handle what gnome-term sends, and correctly terminates children processes.
- And neither can I reproduce the issue in a `kitty bash` session. When exiting kitty, bash seems to handle what kitty sends, and correctly terminates children processes.
- Of course, `gnome-terminal -- bash` does the right thing.

So of the 4 {gnome-term, kitty} x {fish, bash} combinations, only the kitty + fish combination exhibits the issue. Questions at that point:

- Does fish need configuration to be told to handle `SIGHUP` by terminating children processes?
- Is it nonstandard for a terminal emulator to signal exit by sending a `SIGHUP` to the child shell? If it is, well, bash handles it, why not fish?
- Any idea what gnome-term sends on exiting (and that is correctly handled by fish)?
- Anything else coming to mind / any additional questions to ask to kitty's maintainer?

**System info:**

- Problem reproduced in both fish master (`3.0.2-1404-ga33f0eb63` today) and 3.0.2-release
- Kitty 0.14.3
- Linux 5.2.4-arch1
- $TERM = xterm-kitty
- Bug reproduced with a fresh fish without any config

Thanks for fish.

![children-process-alive-after-kitty-exit](https://user-images.githubusercontent.com/522085/62053319-5d276200-b1e5-11e9-9ebf-260265c9df4c.gif)

**Comment by faho** [comment]
>Does fish need configuration to be told to handle SIGHUP by terminating children processes?

No.

>Is it nonstandard for a terminal emulator to signal exit by sending a SIGHUP to the child shell? If it is, well, bash handles it, why not fish?

No, that's also what gnome-terminal does.

----

I can't reproduce this with fish 3.0.2 or git master, and it's supposed to work.

What I'd love to see is `strace`. Start fish, get the pid, start tig, connect strace via `sudo strace -p thatpid`, then kill fish.

For me, I get something like

```output
wait4(-30551, 0x7ffd45868f7c, WSTOPPED, NULL) = ? ERESTARTSYS (To be restarted if SA_RESTART is set) 
# the above is fish waiting on the tig process to end
--- SIGHUP {si_signo=SIGHUP, si_code=SI_KERNEL} ---
--- SIGCONT {si_signo=SIGCONT, si_code=SI_KERNEL} ---
# this is where we got a SIGHUP to hang up, and a SIGCONT in case we were stopped
# ... (a whole bunch more output, mostly writing errors in case this *wasn't* the terminal closing)
getpgrp()                               = 30450
kill(-30551, SIGHUP)                    = 0
# we passed on the HUP to our child's pgroup
# ... (yet more output)
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=30551, si_uid=1000, si_status=0, si_utime=2, si_stime=2} ---
# we got notified that our child died
# ...  (yet more)
```

Bash handles pgroups a bit differently here (the process just stays in its one, which is why it `kill()`s via PID instead of via pgroup), but ours should be working, and appears to be on my system.

So I'd like your `strace` output on gnome-terminal and kitty.

**Comment by ronjouch** [comment]
@faho aaaaaaaaaand the culprit is ... sloppy testing & reporting from me \o/ !

The problem was that, out of habit, I ran my daily `marathon kitty` runner using my own little app runner [marathon](https://github.com/ronjouch/marathon) (a simple `sh`-based runner letting me bind keyboard shortcuts to running or focusing apps), whereas the other 3 cases were run using the regular GNOME Shell Alt+F2 launcher.

Three months ago, I got bored of seeing `sh` in my system monitor as the parent process of all processes started by `marathon`, so I looked for a solution to that, and I botched https://github.com/ronjouch/marathon/commit/504af58ad0b7763b4dfc182c65f118d7a6f1d32c ... which works as intended for the purposes of my script, except NOHUPing kitty had the consequences you know 🙂.

To be frank, I don't really know my shell/terminal fu well, and I still don't understand why `nohup`ing caused this issue. This commit is pure stackoverflow-oriented programming, I just looked for a recipe and applied it. Will research more what `nohup`ing something entails.

Anyway, https://github.com/ronjouch/marathon/commit/d97b09720a6d4d1f5273cfe3adeadb1d3633d942 is probably a better idea, it has all the behavior I need: 1. make `sh` not stay visible in memory and in my process manager, and 2. not cause this bug.

Sorry for the noise, thanks for the fast response, and thanks again for the wonderful work on fish 🙏.
